compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3266,7 +3266,7 @@ object Types {
             TypeComparer.lub(tp1.widenUnionWithoutNull, tp2.widenUnionWithoutNull, canConstrain = true) match
               case union: OrType => union.join
               case res => res
-          else derivedOrType(tp1.widenUnionWithoutNull, tp2.widenUnionWithoutNull)
+          else this
```

Comment:
The documentation for widenUnion would need to be updated with this change since it explicitly talks about widening inside hard unions: https://github.com/lampepfl/dotty/blob/a83a0a7e025d4c248291e7832e38699efa794576/compiler/src/dotty/tools/dotc/core/Types.scala#L1266-L1269